### ISSUE TITLE
1.0 ComponentProvider? (ng1)

### ISSUE THREAD

**Issue by patlux** [angular-ui/ui-router#2995]
Hello,

is there something like a 'ComponentProvider' to give a view dynamically a component like with [ControllerProvider](https://ui-router.github.io/docs/latest/interfaces/ng1.ng1viewdeclaration.html#controllerprovider)?

e.g.:

``` javascript
componentProvider: function(MyResolveData, $transition$) {
  if (MyResolveData.foo) {
    return "Foo"
  } else if ($transition$.to().name === 'bar') {
    return "Bar";
  } else {
    return MyResolveData.component;
  }
}
```

My current workaround is something like this:

``` javascript
templateProvider: function(MyResolveData) {
  return "<" + MyResolveData.component + "></" + MyResolveData.component + ">";
}
```

But with my workaround, it's not possible to use the `Methods` from [Ng1Controller](https://ui-router.github.io/docs/latest/interfaces/ng1.ng1controller.html#uicanexit).

**Comment by christopherthielen** [comment]
There is not, but it seems like there should be.  I'd be open to a PR which implements this.

I'd just need:
- a unit test or three
- docs for `componentProvider`

**Comment by sarahpads** [comment]
Hi guys,

I'm looking at how this feature could be implemented.  I'm new to the code base and would really appreciate any guidance.

Am I correct in understanding that the component logic would have to be moved from the ng1ViewsBuilder (statebuilders/views.ts:24) into the Ng1ViewConfig class definition (statebuilders/views.ts:123)?  I came to this understanding because the PathNode information is available in the class definition.

Am I completely off the mark?

**Comment by christopherthielen** [comment]
@sarahpads happy to provide any insights I can.

I think the block of code after this comment:
`// Dynamically build a template like "<component-name input1='::$resolve.foo'></component-name>"`
probably belongs in [template factory](https://github.com/angular-ui/ui-router/blob/master/src/ng1/templateFactory.ts).

It could be named `fromComponent`, and then have a provider version called `fromComponentProvider`.

The current `fromProvider` function shows how to inject a provider function with the proper resolve context (this make sure it can inject the right `$transition$`, etc).  That function can probably be reused, extracted as a standalone or static fn.

**Comment by sarahpads** [comment]
@christopherthielen I've run into something of a roadblock.

In the templateFactory fromConfig method, config.componentProvider is always undefined.  I'm initializing my state like this in stateSpec:
 ```.state('dynamicComponent', {
        url: '/dynamicComponent/:type',
        componentProvider: ['$stateParams', function($stateParams) {
          componentName = $stateParams.type + 'Component';
          return componentName;
        }]
      })```

Is it possible that the 'componentProvider' attribute is getting stripped out during the state registration process?  Would you be able to point me in the right direction?

Thanks!

**Comment by sarahpads** [comment]
Hi again,

I have put together a pull request that addresses this thread: https://github.com/angular-ui/ui-router/pull/3165.

Please let me know if there are any improvements I can make or if I've stomped over some coding conventions.

Thanks!
Sarah

**Comment by mackelito** [comment]
What are the benefits of routing to the component directly vs using a template like "<component-name input1='::$resolve.foo'></component-name>" ?

**Comment by christopherthielen** [comment]
@mackelito it's just sugar for building the template (like your example).  It can reduce boilerplate, and follows the same mental model as the other component based ui-routers (ng2 and react)

In the future there may be other benefits for ng1, such as propagating callback bindings from the parent ui-view tag to the child component.

**Comment by christopherthielen** [comment]
Closed by 4b583603d7cb63ef1cd36728763e2b028934a63f / #3165
